src/common/NetworkManager.ts

```diff
@@ -399,28 +512,83 @@ export class NetworkManager extends EventEmitter {
     this.emit(NetworkManagerEmittedEvents.RequestServedFromCache, request);
   }
 
-  _handleRequestRedirect(
-    request: HTTPRequest,
-    responsePayload: Protocol.Network.Response
+  _emitResponseEvent(
+    responseReceived: Protocol.Network.ResponseReceivedEvent,
+    extraInfo: Protocol.Network.ResponseReceivedExtraInfoEvent
   ): void {
-    const response = new HTTPResponse(this._client, request, responsePayload);
-    request._response = response;
-    request._redirectChain.push(request);
-    response._resolveBody(
-      new Error('Response body is unavailable for redirect responses')
+    const request = this._requestIdToRequest.get(responseReceived.requestId);
+    // FileUpload sends a response without a matching request.
+    if (!request) return;
+
+    const extraInfos = this._requestIdToResponseExtraInfo(
+      responseReceived.requestId
     );
-    this._forgetRequest(request, false);
+    if (extraInfos.length) {
```

Comment:
@josepharhar in particular, I have added this check here: is my understanding correct that there should be no unprocessed extraInfo events if we emit _emitResponseEvent?